github.com/cockroachdb/cockroach/pkg/sql: TestLeaseManagerDrain failed under stress

SHA: https://github.com/cockroachdb/cockroach/commits/6dbac80561b565d2f5e8304b9f36c4ed9112f0f2

Parameters:
```
COCKROACH_PROPOSER_EVALUATED_KV=true
TAGS=deadlock
GOFLAGS=
```

Stress build found a failed test: https://teamcity.cockroachdb.com/viewLog.html?buildId=117241&tab=buildLog

```
W170113 09:54:22.304804 2279390 server/status/runtime.go:116  Could not parse build timestamp: parsing time "" as "2006/01/02 15:04:05": cannot parse "" as "2006"
I170113 09:54:22.306996 2279390 server/config.go:456  1 storage engine initialized
I170113 09:54:22.308869 2279390 server/node.go:426  [n?] store [n0,s0] not bootstrapped
I170113 09:54:22.316763 2279390 server/node.go:355  [n?] **** cluster 2f84a33a-c847-4a7e-96da-46dc90a686bb has been created
I170113 09:54:22.316848 2279390 server/node.go:356  [n?] **** add additional nodes by specifying --join=127.0.0.1:35269
I170113 09:54:22.319644 2279390 storage/store.go:1250  [n1] [n1,s1]: failed initial metrics computation: [n1,s1]: system config not yet available
I170113 09:54:22.319805 2279390 server/node.go:439  [n1] initialized store [n1,s1]: {Capacity:536870912 Available:536870912 RangeCount:1 LeaseCount:0}
I170113 09:54:22.319927 2279390 server/node.go:324  [n1] node ID 1 initialized
I170113 09:54:22.320024 2279390 gossip/gossip.go:292  [n1] NodeDescriptor set to node_id:1 address:<network_field:"tcp" address_field:"127.0.0.1:35269" > attrs:<> locality:<> 
I170113 09:54:22.320262 2279390 storage/stores.go:296  [n1] read 0 node addresses from persistent storage
I170113 09:54:22.320366 2279390 server/node.go:569  [n1] connecting to gossip network to verify cluster ID...
I170113 09:54:22.321160 2279390 server/node.go:589  [n1] node connected via gossip and verified as part of cluster "2f84a33a-c847-4a7e-96da-46dc90a686bb"
I170113 09:54:22.321232 2279390 server/node.go:374  [n1] node=1: started with [[]=] engine(s) and attributes []
I170113 09:54:22.321300 2279390 sql/executor.go:322  [n1] creating distSQLPlanner with address {tcp 127.0.0.1:35269}
I170113 09:54:22.323056 2279390 server/server.go:621  [n1] starting https server at 127.0.0.1:39457
I170113 09:54:22.323111 2279390 server/server.go:622  [n1] starting grpc/postgres server at 127.0.0.1:35269
I170113 09:54:22.323162 2279390 server/server.go:623  [n1] advertising CockroachDB node at 127.0.0.1:35269
I170113 09:54:22.328352 2285294 storage/split_queue.go:99  [split,n1,s1,r1/1:/M{in-ax},@c43d9d6f00] splitting at keys [/Table/11/0 /Table/12/0 /Table/13/0 /Table/14/0]
I170113 09:54:22.377229 2285325 sql/event_log.go:95  [n1] Event: "node_join", target: 1, info: {Descriptor:{NodeID:1 Address:{NetworkField:tcp AddressField:127.0.0.1:35269} Attrs: Locality:} ClusterID:2f84a33a-c847-4a7e-96da-46dc90a686bb StartedAt:1484301262321203479}
I170113 09:54:22.387503 2285294 storage/replica_command.go:2354  [split,n1,s1,r1/1:/M{in-ax},@c43d9d6f00] initiating a split of this range at key /Table/11 [r2]
E170113 09:54:22.415251 2285294 storage/queue.go:599  [split,n1,s1,r1/1:/{Min-Table/11},@c43d9d6f00] unable to split [n1,s1,r1/1:/{Min-Table/11}] at key "/Table/12/0": key range /Table/12/0-/Table/12/0 outside of bounds of range /Min-/Max
I170113 09:54:22.416838 2285294 storage/split_queue.go:99  [split,n1,s1,r2/1:/{Table/11-Max},@c438fc6000] splitting at keys [/Table/12/0 /Table/13/0 /Table/14/0]
I170113 09:54:22.416970 2285294 storage/replica_command.go:2354  [split,n1,s1,r2/1:/{Table/11-Max},@c438fc6000] initiating a split of this range at key /Table/12 [r3]
I170113 09:54:22.429159 2279390 sql/event_log.go:95  [n1] Event: "alter_table", target: 12, info: {TableName:eventlog Statement:ALTER TABLE system.eventlog ALTER COLUMN uniqueID SET DEFAULT uuid_v4() User:node MutationID:0 CascadeDroppedViews:[]}
E170113 09:54:22.440887 2285294 storage/queue.go:599  [split,n1,s1,r2/1:/Table/1{1-2},@c438fc6000] unable to split [n1,s1,r2/1:/Table/1{1-2}] at key "/Table/13/0": key range /Table/13/0-/Table/13/0 outside of bounds of range /Table/11-/Max
I170113 09:54:22.442750 2285294 storage/split_queue.go:99  [split,n1,s1,r3/1:/{Table/12-Max},@c43c2c4900] splitting at keys [/Table/13/0 /Table/14/0]
I170113 09:54:22.443011 2285294 storage/replica_command.go:2354  [split,n1,s1,r3/1:/{Table/12-Max},@c43c2c4900] initiating a split of this range at key /Table/13 [r4]
E170113 09:54:22.503388 2285294 storage/queue.go:599  [split,n1,s1,r3/1:/Table/1{2-3},@c43c2c4900] unable to split [n1,s1,r3/1:/Table/1{2-3}] at key "/Table/14/0": key range /Table/14/0-/Table/14/0 outside of bounds of range /Table/12-/Max
I170113 09:54:22.503979 2285294 storage/split_queue.go:99  [split,n1,s1,r4/1:/{Table/13-Max},@c43c01b200] splitting at keys [/Table/14/0]
I170113 09:54:22.504079 2285294 storage/replica_command.go:2354  [split,n1,s1,r4/1:/{Table/13-Max},@c43c01b200] initiating a split of this range at key /Table/14 [r5]
I170113 09:54:22.535958 2279390 server/server.go:676  [n1] done ensuring all necessary migrations have run
I170113 09:54:22.536044 2279390 server/server.go:678  [n1] serving sql connections
POTENTIAL DEADLOCK: Inconsistent locking. saw this ordering in one goroutine:
happened before
lease.go:714 sql.(*tableState).release { t.mu.Lock() } <<<<<
lease.go:1153 sql.(*LeaseManager).Release { return t.release(lease, m) }
lease_test.go:133 sql_test.(*leaseTest).release { return t.node(nodeID).Release(lease) }
lease_test.go:147 sql_test.(*leaseTest).mustRelease { if err := t.release(nodeID, lease); err != nil { }
lease_test.go:427 sql_test.TestLeaseManagerDrain { t.mustRelease(1, l1, nil) }

happened after
lease.go:1157 sql.(*LeaseManager).isDraining { m.mu.Lock() } <<<<<
lease.go:732 sql.(*tableState).release.func1 { m.isDraining() || }
lease.go:751 sql.(*tableState).release { if decRefcount(s) { }
lease.go:1153 sql.(*LeaseManager).Release { return t.release(lease, m) }
lease_test.go:133 sql_test.(*leaseTest).release { return t.node(nodeID).Release(lease) }
lease_test.go:147 sql_test.(*leaseTest).mustRelease { if err := t.release(nodeID, lease); err != nil { }
lease_test.go:427 sql_test.TestLeaseManagerDrain { t.mustRelease(1, l1, nil) }

in another goroutine: happened before
lease.go:1165 sql.(*LeaseManager).SetDraining { m.mu.Lock() } <<<<<
lease_test.go:434 sql_test.TestLeaseManagerDrain { t.nodes[1].SetDraining(true) }

happend after
lease.go:1172 sql.(*LeaseManager).SetDraining { t.mu.Lock() } <<<<<
lease_test.go:434 sql_test.TestLeaseManagerDrain { t.nodes[1].SetDraining(true) }


Other goroutines holding locks:
goroutine 2279390 lock 0xc42b457328
lease.go:1165 sql.(*LeaseManager).SetDraining { m.mu.Lock() } <<<<<
lease_test.go:434 sql_test.TestLeaseManagerDrain { t.nodes[1].SetDraining(true) }




ERROR: exit status 2

make: *** [stress] Error 1
1 runs completed, 1 failures, over 56s
Makefile:147: recipe for target 'stress' failed
```

--- a/pkg/sql/lease.go
+++ b/pkg/sql/lease.go
@@ -22,6 +22,7 @@ import (
 	"fmt"
 	"math/rand"
 	"sort"
+	"sync/atomic"
 	"time"
 
 	"golang.org/x/net/context"
@@ -956,9 +957,12 @@ func makeTableNameCacheKey(dbID sqlbase.ID, tableName string) tableNameCacheKey 
 // LeaseManager.mu > tableState.mu > tableNameCache.mu > LeaseState.mu
 type LeaseManager struct {
 	LeaseStore
-	mu       syncutil.Mutex
-	draining bool
-	tables   map[sqlbase.ID]*tableState
+	mu struct {
+		syncutil.Mutex
+		tables map[sqlbase.ID]*tableState
+	}
+
+	draining atomic.Value
 
 	// tableNames is a cache for name -> id mappings. A mapping for the cache
 	// should only be used if we currently have an active lease on the respective
@@ -988,13 +992,18 @@ func NewLeaseManager(
 			testingKnobs: testingKnobs.LeaseStoreTestingKnobs,
 			memMetrics:   memMetrics,
 		},
-		tables:       make(map[sqlbase.ID]*tableState),
 		testingKnobs: testingKnobs,
 		tableNames: tableNameCache{
 			tables: make(map[tableNameCacheKey]*LeaseState),
 		},
 		stopper: stopper,
 	}
+
+	lm.mu.Lock()
+	lm.mu.tables = make(map[sqlbase.ID]*tableState)
+	lm.mu.Unlock()
+
+	lm.draining.Store(false)
 	return lm
 }
 
@@ -1154,21 +1163,19 @@ func (m *LeaseManager) Release(lease *LeaseState) error {
 }
 
 func (m *LeaseManager) isDraining() bool {
-	m.mu.Lock()
-	defer m.mu.Unlock()
-	return m.draining
+	return m.draining.Load().(bool)
 }
 
 // SetDraining (when called with 'true') removes all inactive leases. Any leases
 // that are active will be removed once the lease's reference count drops to 0.
 func (m *LeaseManager) SetDraining(drain bool) {
-	m.mu.Lock()
-	defer m.mu.Unlock()
-	m.draining = drain
+	m.draining.Store(drain)
 	if !drain {
 		return
 	}
-	for _, t := range m.tables {
+	m.mu.Lock()
+	defer m.mu.Unlock()
+	for _, t := range m.mu.tables {
 		t.mu.Lock()
 		t.releaseInactiveLeases(m.LeaseStore)
 		t.mu.Unlock()
@@ -1179,10 +1186,10 @@ func (m *LeaseManager) SetDraining(drain bool) {
 func (m *LeaseManager) findTableState(tableID sqlbase.ID, create bool) *tableState {
 	m.mu.Lock()
 	defer m.mu.Unlock()
-	t := m.tables[tableID]
+	t := m.mu.tables[tableID]
 	if t == nil && create {
 		t = &tableState{id: tableID, tableNameCache: &m.tableNames, stopper: m.stopper}
-		m.tables[tableID] = t
+		m.mu.tables[tableID] = t
 	}
 	return t
 }

